Render OTP2 routing error messages (DT-5405)

joonaojapalo opened the pull request:
## Proposed Changes

  - Rendering of OTP2 error messages based on the design (https://www.figma.com/file/dXPBr0Ow9BcOEi4hXqeca5/Summary?node-id=1401%3A3745) 
  - Refactoring of `ItinerarySmmaryList` to subcomponents in new subdirectory.
  - Unit tests for error message matching login.
  - Improved package.json scripts to enable more efficient test driven approach. Eg. `yarn test-unit:app -g 'ComponentName' -- --watch` to watch and run of single test case.

Related issue: DT-5405

## Pull Request Check List

  - A reasonable set of unit tests is included
  - Console does not show new warnings/errors
  - Changes are documented or they are self explanatory
  - This pull request does not have any merge conflicts
  - All existing tests pass in CI build
  - Code coverage does not decrease (unless measured incorrectly)

## Review

  - Read and verify the code changes
  - Test the functionality by running the UI locally with all popular browsers available in your platform
  - Check that the implementation matches the design, when such one is defined in a Jira issue
  - Merge the pull request

Commit:
chore: prettify README

Commit:
chore: streamline running unit tests for

Commit:
chore: clarify e2e visual test case setup

Commit:
fix: add OTP (v2) messageEnum strings

Commit:
feat: add routing error support for OTP2 errors

Commit:
feat: add translations

Commit:
chore: add unit tests for ItinerarySummaryMessage

Commit:
chore: move RoutingErrorShape to prop-types

Commit:
feat: sv translations to router-* keys

joonaojapalo commented:
All required translations are included.

optionsome committed:
Merge branch 'v3' into DT-5405

optionsome commented on README.md:
```diff
@@ -1,39 +1,69 @@
 
 More information about [testing](docs/Tests.md).
 
+### Unit testing
```
There is similar text already in Tests.md file. Should that be updated instead/in addition to this?

optionsome commented on app/component/ItinerarySummaryListContainer/ItinerarySummaryListContainer.js:
```diff
@@ -2,22 +2,27 @@ import PropTypes from 'prop-types';
 import React, { useState } from 'react';
```
Is there some best practice for putting these components into directories? This structure hasn't been used before in this project (or only once in some recent update). I like that the files are grouped together. Only questions I have: is it necessary/good to have the index.js in each directory? Why is there a nested component structure? Is it because a component uses some other component which then uses some components? Some of the files in the components directory are more like util classes rather than components, should there be some different directory for those?

joonaojapalo replied:
Good points!

In this PR, `index.js` file keeps the import statement untouched.

Splitting big component to smaller components with more strict responsibilities yields many components and I think it's natural to store these hierarchially rather than as a flat list and let the directory layout help with code-base navigation. Eg. SummaryPage, apporx. 2400 LOC probably could be splitted to dozens of child components. In hsl.fi project there's a good example of using nesting: [CityBikesRentalsBlock](https://dev.azure.com/hslfi/hslfi/_git/hslfi-frontend?path=%2FFrontend%2Fsrc%2Fblocks%2FCityBikesRentalsBlock&version=GBdevelop&_a=contents) (they seem to use one index.ts file per directory). This nested hierarchy definitely needs some rules simplify code-base navigation rather than making it more complex. Natural basic rule is that components should only import components from *subdirectory or from parent directories*, but *not from any parent's subdirectory*. This way generic components would live higher in the directory hierarchy. Also it would enable extracting a child component without need to worry about adding noise/complexity to directory listings.

I think that classification simple and pure utility components from more fat container like orchestrator components boils down to a debated question of how to differentiate [presentational vs container components](https://medium.com/@dan_abramov/smart-and-dumb-components-7ca2f9a7c7d0). It seems that the current practice is not to make this separation and just treat all components equally.

Just wondering should we run a single automated batch operation to convert all current `app/components/Xxx.js` to `app/component/Xxx/{Xxx.js,index.js}` structure to make project layout homogenous?

optionsome replied:
I'm not sure if the file `findErrorMessageIds.js` is either presentational or container component as it just contains utility functions but no react.

I think we first try use this directory structure for a bit and see if we find any issues with it, if not, then we could do the refactoring to change all the components to be in directories.

joonaojapalo replied:
Indeed, I think `utils` subdirectory will work fine. 👍  I read the comment too quickly and not properly checked the exact files in question, excuse my misunderstanding.

optionsome commented on app/component/ItinerarySummaryListContainer/components/components/Link.js:
```diff
@@ -0,0 +1,19 @@
+import { FormattedMessage } from 'react-intl';
+import cx from 'classnames';
+
+const Link = ({ msgId, href }) => (
```
We already use the Link component from found in a lot of places which is meant for links within this site. Should we name this something like ExternalLink so the difference between this and the found Link is more obvious. I was quite surprised there wasn't already a component like this somewhere. I would assume something like this would also be needed in other places. Is putting this component inside this ItinerarySummaryListContainer directory the right place? Is this generic enough so it could also be used elsewhere?

joonaojapalo replied:
Actually there already exists the component https://github.com/HSLdevcom/digitransit-ui/blob/v2/app/component/ExternalLink.js but it has a different DOM structure and CSS definitions which affect positioning. There are 18 external anchor tags in the code base (`<a>`) but almost all of them have individual styling.  I think it's better to substitute this back to `ChangeDepartureTimeLink` and `PastLink`.

optionsome commented on app/component/ItinerarySummaryListContainer/components/errorCardProperties.js:
```diff
@@ -0,0 +1,205 @@
+    },
+  },
+  {
+    id: 'walk-bike-itinerary-4',
```
why is 4 before 1 in this list?

joonaojapalo replied:
This directly from priority definition: https://github.com/HSLdevcom/digitransit-ui/blob/v2/app/component/ItinerarySummaryListContainer.js#L263

optionsome commented on app/translations.js:
```diff
@@ -1120,7 +1120,7 @@ const translations = {
     'itinerary-in-the-past':
       'The route search falls within a period that is in the past.',
-    'itinerary-in-the-past-link': 'I want to travel now ›',
+    'itinerary-in-the-past-link': 'Depart now ›',
```
Why does this have the arrow in the translation while some others don't even though it should also be shown with "Change departure time", for example?

joonaojapalo replied:
Carets are indeed missing from some translation strings - now added.

optionsome reviewed (commented):
Haven't tested this properly yet, should read through the code. I like that the big components are split into smaller pieces. I merged the current v3 into this and pushed it.

Commit:
chore: format docs/Tests.md

Commit:
chore: remove duplicate unit test docs

Commit:
chore: rename ItinerarySummaryMessage/Link to ExternalLink

Commit:
chore: substitute ItinerarySummaryListContainer/ExternalLink back

Commit:
fix: add missing caret to translations

optionsome commented:
Seems like some translations need to be updated. At least for NO_TRANSIT_CONNECTION the swedish translations were incorrect

Commit:
fix: add missing sv translations

Commit:
chore: move errorCardProperties.js and findErrorMessageIds.js to utils/

optionsome reviewed (commented):
NO_TRANSIT_CONNETION and NO_TRANSIT_CONNECTION_IN_SEARCH_WINDOW swedish translation had wrong word in title: ruttförslag -> ruttalternativ.

NO_TRANSIT_CONNETION and NO_TRANSIT_CONNECTION_IN_SEARCH_WINDOW finnish title had "." at the end of the title even though it shouldn't.

Compared to production reittiopas, the OUTSIDE_BOUNDS type errors are now rarely returned. I think it makes more sense to return those when we are outside of the service area than the LOCATION_NOT_FOUND type errors.

I think we could display the system error message when routing returns 5xx error or GraphQL error that starts with "Exception while fetching data" (if there are no itineraries returned).

Commit:
fix: translations

Commit:
Merge branch 'v3' into DT-5405

Commit:
fix: render system-error on routing service http errors + detect outside-bounds client-side

Commit:
chore: freeze object constant

Commit:
chore: package.json - join common scripts

joonaojapalo commented:
> NO_TRANSIT_CONNETION and NO_TRANSIT_CONNECTION_IN_SEARCH_WINDOW swedish translation had wrong word in title: ruttförslag -> ruttalternativ.
>
> NO_TRANSIT_CONNETION and NO_TRANSIT_CONNECTION_IN_SEARCH_WINDOW finnish title had "." at the end of the title even though it shouldn't.

Fixed!

> Compared to production reittiopas, the OUTSIDE_BOUNDS type errors are now rarely returned. I think it makes more sense to return those when we are outside of the service area than the LOCATION_NOT_FOUND type errors.

Fixed! Added origin and destination point client-side validation against `config.areaPolygon` property.

> I think we could display the system error message when routing returns 5xx error or GraphQL error that starts with "Exception while fetching data" (if there are no itineraries returned).

Verified that SYSTEM_ERROR is given on HTTP error status. However, it seems that in case GraphQL server responds with partially sensible payload but contains an error, `react-relay` doesn't propagate payload to wrapped component even though [specification should allow this](https://spec.graphql.org/June2018/#sec-Errors). I think this behavior could be tolerated.

optionsome commented:
Noticed a few thing while retesting this.
* At least the system error was displayed even though some itineraries were returned. I have some odd memory that maybe the props were not passed to the to components if there was an error but maybe it was just when the request was sent as part of a found `Route`. Hopefully we can render the result even if there are errors. I tested this running otp locally where I had edited https://github.com/opentripplanner/OpenTripPlanner/blob/dev-2.x/src/ext/java/org/opentripplanner/ext/legacygraphqlapi/datafetchers/LegacyGraphQLItineraryImpl.java#L73-L76 to be like:
```@Override
  public DataFetcher<Long> startTime() {
    return environment -> {
      if (getSource(environment).startTime().toInstant().toEpochMilli() > 0) {
        throw new RuntimeException();
      }
      return getSource(environment).startTime().toInstant().toEpochMilli();
    };
  }
```
* At least the OUTSIDE_SERVICE_PERIOD causes some rendering issues in mobile. It just renders an empty error box.
* We could change the links to opas.matka.fi include the user's language like opas.matka.fi/sv/ or opas.matka.fi/en/

Commit:
fix: i18n nationalService link

Commit:
fix: routeErrors on mobile view

optionsome commented on app/configurations/config.hsl.js:
```diff
@@ -117,7 +117,20 @@ export default {
+    },
+    sv: {
+      name: 'matka.fi',
+      href: 'https://opas.matka.fi/?locale=sv',
```
```suggestion
      href: 'https://opas.matka.fi/sv/',
```

optionsome commented on app/configurations/config.hsl.js:
```diff
@@ -117,7 +117,20 @@ export default {
+    },
+    en: {
+      name: 'matka.fi',
+      href: 'https://opas.matka.fi/?locale=en',
```
```suggestion
      href: 'https://opas.matka.fi/en/',
```

optionsome commented on app/configurations/config.waltti.js:
```diff
@@ -126,7 +126,20 @@ export default {
+    },
+    sv: {
+      name: 'matka.fi',
+      href: 'https://opas.matka.fi/?locale=sv',
```
```suggestion
      href: 'https://opas.matka.fi/sv/',
```

optionsome commented on app/configurations/config.waltti.js:
```diff
@@ -126,7 +126,20 @@ export default {
+    },
+    en: {
+      name: 'matka.fi',
+      href: 'https://opas.matka.fi/?locale=en',
```
```suggestion
      href: 'https://opas.matka.fi/en/',
```

optionsome commented on test/unit/component/ItinerarySummaryMessage.test.js:
```diff
@@ -0,0 +1,322 @@
+    },
+    sv: {
+      name: 'matka.fi',
+      href: 'https://opas.matka.fi/?locale=sv',
```
```suggestion
      href: 'https://opas.matka.fi/sv/',
```

optionsome commented on test/unit/component/ItinerarySummaryMessage.test.js:
```diff
@@ -0,0 +1,322 @@
+    },
+    en: {
+      name: 'matka.fi',
+      href: 'https://opas.matka.fi/?locale=en',
```
```suggestion
      href: 'https://opas.matka.fi/en/',
```

optionsome reviewed (commented):
If I remember correctly, there was some slight difference between linking with locale param vs the `/<lang>/` url path and I think the path was better options. Tested and everything seems to work nicely now.

joonaojapalo committed:
Update app/configurations/config.waltti.js

Co-authored-by: Joel Lappalainen <joel.lappalainen@cgi.com>

joonaojapalo committed:
Update app/configurations/config.waltti.js

Co-authored-by: Joel Lappalainen <joel.lappalainen@cgi.com>

joonaojapalo committed:
Update test/unit/component/ItinerarySummaryMessage.test.js

Co-authored-by: Joel Lappalainen <joel.lappalainen@cgi.com>

joonaojapalo committed:
Update test/unit/component/ItinerarySummaryMessage.test.js

Co-authored-by: Joel Lappalainen <joel.lappalainen@cgi.com>

joonaojapalo committed:
Update app/configurations/config.hsl.js

Co-authored-by: Joel Lappalainen <joel.lappalainen@cgi.com>

joonaojapalo committed:
Update app/configurations/config.hsl.js

Co-authored-by: Joel Lappalainen <joel.lappalainen@cgi.com>
